examples/generative/ddim.py

```diff
@@ -0,0 +1,839 @@
+build a [U-Net](https://arxiv.org/abs/1505.04597) with identical input and output
+dimensions. U-Net is a popular semantic segmentation architecture, whose main idea is
+that it progressively downsamples and then upsamples its input image, and adds skip
+connections between layers having the same resolution. These help with gradient flow and
+avoid introducing a representation bottleneck, unlike usual
+[autoencoders](https://www.deeplearningbook.org/contents/autoencoders.html). Based on
+this, one can view
+[diffusion models as denoising autoencoders](https://benanne.github.io/2022/01/31/diffusion.html)
+without a bottleneck.
+
+The network takes two inputs, the noisy images and the variances of their noise
+components. The latter is required since denoising a signal requires different operations
+at different levels of noise. We transform the noise variances using sinusoidal
+embeddings, similarly to positional encodings used both in
+[transformers](https://arxiv.org/abs/1706.03762) and
+[NeRF](https://arxiv.org/abs/2003.08934). This helps the network to be
+[highly sensitive](https://arxiv.org/abs/2006.10739) to the noise level, which is
+crucial for good performance. We implement sinusoidal embeddings using a
+[Lambda layer](https://keras.io/api/layers/core_layers/lambda/).
+
+Some other considerations:
+
+* We build the network using the
+[Keras Functional API](https://keras.io/guides/functional_api/), and use
+[closures](https://twitter.com/fchollet/status/1441927912836321280) to build blocks of
+layers in a consistent style.
+* [Diffusion models](https://arxiv.org/abs/2006.11239) embed the index of the timestep of
+the diffusion process instead of the noise variance, while
+[score-based models (Table 1)](https://arxiv.org/abs/2206.00364)
+usually use some function of the noise level. I
+prefer the latter so that we can change the sampling schedule at inference time, without
+retraining the network.
+* [Diffusion models](https://arxiv.org/abs/2006.11239) input the embedding to each
+convolution block separately. We only input it at the start of the network for
+simplicity, which in my experience barely decreases preformance, because the skip and
+residual connections help the information propagate through the network properly.
+* In the literature it is common to use
+[attention layers](https://keras.io/api/layers/attention_layers/multi_head_attention/)
+at lower resolutions for better global coherence. I omitted it for simplicity.
+* We disable the learnable center and scale parameters of the batch normalization layers,
+since the following convolution layers make them redundant.
+* We initialize the last convolution's kernel to all zeros as a good practice, making the
+network predict only zeros after initialization, which is the mean of its targets. This
+will improve behaviour at the start of training and make the mean squared error loss
+start at exactly 1.
+"""
+
+
+def get_network():
```

Comment:
Following the closure idea, it would be more consistent to give this is a capitalized name

Reply:
Fair, but this returns a model object, not a function like the closures.